### DISCUSSION TITLE
post/way-to-functional-programming/

### DISCUSSION THREAD

**Discussion by giscus** [kakaopay/techlog-comments#106] [Announcements]
# post/way-to-functional-programming/

함수형 프로그래밍 막막하셨나요? 제가 먼저 찾아서 가본 길을 따라와 보실래요?

https://tech.kakaopay.com/post/way-to-functional-programming/

**Comment by Denia-park** [comment]
안녕하세요. 좋은 글 잘 봤습니다.

글을 읽고 따라해보다가 컴파일 에러가 발생하여 한가지 여쭤보고 싶은 사항이 있습니다.

`3. 작은 패턴의 발견과 추상화` 에서 알려주신 예제코드를 실제로 작성하면 
`override fun <B> recover(recoverFunc: (ResultCode) -> SampleResult<B>): SampleResult<B> = this`
해당 부분에서 컴파일 에러가 발생합니다!

컴파일 에러 내용은 타입이 안 맞아서 `this`를 사용할 수 없다고 나옵니다.

해당 부분을 어떻게 해결하셨을까요 ??

//전체 코드
``` kotlin
    data class Success<out A>(val value: A) : SampleResult<A>() {
        override fun <B> map(f: (A) -> B): SampleResult<B> = Success(f(value))
        override fun <B> flatMap(f: (A) -> SampleResult<B>): SampleResult<B> = f(value)
        override fun <B> recover(recoverFunc: (ResultCode) -> SampleResult<B>): SampleResult<B> = this
        override fun toString(): String = "SUCCESS"
    }
```
